### ISSUE TITLE
Function overloading eslint(no-unused-vars) error.

### ISSUE THREAD

**Issue by giffeler** [typescript-eslint/typescript-eslint#2859]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
function test (id: number): void
function test (id: string): void
function test (id: any): void { console.log(id); }
```

```tsconfig
	"compilerOptions": {
		"allowJs": false,
		"allowSyntheticDefaultImports": true,
		"alwaysStrict": true,
		"charset": "utf8",
		"checkJs": false,
		"diagnostics": true,
		"downlevelIteration": false,
		"esModuleInterop": true,
		"forceConsistentCasingInFileNames": true,
		"isolatedModules": false,
		"jsx": "preserve",
		"lib": ["dom", "ES2020"],
		"listEmittedFiles": true,
		"moduleResolution": "node",
		"noEmit": false,
		"noFallthroughCasesInSwitch": true,
		"noImplicitAny": true,
		"noImplicitReturns": true,
		"noImplicitThis": true,
		"noUnusedLocals": true,
		"noUnusedParameters": true,
		"outFile": "niko.js",
		"preserveConstEnums": true,
		"removeComments": true,
		"skipLibCheck": false,
		"sourceMap": true,
		"strict": true,
		"strictBindCallApply": true,
		"strictFunctionTypes": true,
		"strictNullChecks": true,
		"strictPropertyInitialization": true,
		"target": "ES2020"
```
**Expected Result**

No unused vars warning.

**Actual Result**

_'id' is defined but never used. eslint(no-unused-vars)_

**Additional Info**

<img width="526" alt="Screenshot 2020-12-09 at 07 08 39" src="https://user-images.githubusercontent.com/19636707/101592323-58380c80-39ee-11eb-8da0-d70309f653ab.png">

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.9.1` |
| `TypeScript`                | `4.2.0-dev.20201207` |
| `ESLint`                    | `7.15.0` |
| `node`                      | `15.3.0` |

**Comment by bradzacher** [comment]
This is explained in the first FAQ answer - which you ticked the box saying you'd read.

https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code
